pkg/types/azure/validation/machinepool.go

```diff
@@ -124,16 +124,43 @@ func ValidateMachinePool(p *azure.MachinePool, poolName string, platform *azure.
 func validateDataDiskSetup(azurePool *azure.MachinePool, pool *types.MachinePool, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 
-	for _, d := range azurePool.DataDisks {
+	// We could have a situation where the azure DataDisks are
+	// defined but no corresponding disk setup but we should never have
+	// more DiskSetup than DataDisks
+	if len(azurePool.DataDisks) < len(pool.DiskSetup) {
+		allErrs = append(allErrs, field.TooLong(fldPath, pool.DiskSetup, len(azurePool.DataDisks)))
+		// return early if disksetup and datadisks don't match lengths
+		return allErrs
+	}
+
+	for i, d := range azurePool.DataDisks {
 		if d.Lun == nil {
 			allErrs = append(allErrs, field.Required(fldPath.Child("Lun"), fmt.Sprintf("%s must have lun id", d.NameSuffix)))
 		} else if *(d.Lun) < 0 || *(d.Lun) > 63 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("Lun"), fmt.Sprintf("%s must have lun id between 0 and 63", d.NameSuffix)))
 		}
 
-		if d.DiskSizeGB == 0 {
+		if d.DiskSizeGB <= 0 {
 			allErrs = append(allErrs, field.Invalid(fldPath.Child("DiskSizeGB"), d.DiskSizeGB, "diskSizeGB must be greater than zero"))
 		}
+
+		if i < len(pool.DiskSetup) {
+			setup := pool.DiskSetup[i]
+			switch setup.Type {
+			case types.Etcd:
+				if setup.Etcd != nil && setup.Etcd.PlatformDiskID != d.NameSuffix {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("NameSuffix"), d.NameSuffix, fmt.Sprintf("does not match etcd PlatformDiskID %s", setup.Etcd.PlatformDiskID)))
```

Comment:
What about adding comma to values?
```suggestion
					allErrs = append(allErrs, field.Invalid(fldPath.Child("NameSuffix"), d.NameSuffix, fmt.Sprintf("does not match etcd PlatformDiskID %q", setup.Etcd.PlatformDiskID)))
```